### DISCUSSION TITLE
Feature parity with `worker-loader`

### DISCUSSION THREAD

**Discussion by mmazzarolo** [webpack/webpack#14198] [Q&A] [answered]
👋 !
The [`worker-loader`](https://github.com/webpack-contrib/worker-loader) package is now officially deprecated in favor of [Webpack 5 built-in web worker loading strategy](https://webpack.js.org/guides/web-workers/).  

I'm wondering: what are the plans around supporting `worker-loader`'s configuration options in Webpack 5 (if any)? 

From my understanding, there are two main features missing:
- Support inlining web workers (that I assume is covered in [Inline worker in webpack 5](https://github.com/webpack/webpack/discussions/14066)?).  
- Support using `publicPath` for the worker bundle. I recently wrote about this [in a blog post](https://mmazzarolo.com/blog/2021-09-03-loading-web-workers-using-webpack-5/). Am I right assuming the "new" way to deal with it is by setting `__webpack_public_path__ ` at runtime?  

Thanks!

**Comment by alexander-akait** [comment, marked as answer]
> Support inlining web workers (that I assume is covered in Inline worker in webpack 5?).

I don't think we need this in 2021, anyway it is possible to using `assets/inline`

> Support using publicPath for the worker bundle. I recently wrote about this in a blog post. Am I right assuming the "new" way to deal with it is by setting __webpack_public_path__ at runtime?

You can move worker to own entry and setup publicPath or use multi compiler mode

**Reply by mmazzarolo** [reply]
Thanks. 

> I don't think we need this in 2021, anyway it is possible to using assets/inline

Could you add some more context to this or point me to the docs? Do you mean by using `asset/inline` in a module rule that targets just the worker?

**Reply by alexander-akait** [reply]
You can write something like:
```js
try {
  new Worker(new URL('./my.worker.js', import.meta.url));
} catch (error) {
  const URL = new URL('./my.worker.js?fallback', import.meta.url);
  
  // Logic for data/inline
}
```

And register `asset/inline` type for `resourceQuery: /fallback/` like here https://webpack.js.org/guides/asset-modules/#replacing-inline-loader-syntax

**Reply by AprilArcus** [reply]
I followed these instructions and received `Uncaught ReferenceError: require is not defined`.

The base64 encoded script starts with
```
"use strict";

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.symbol.js");

require("core-js/modules/es.symbol.description.js");
```

**Reply by AprilArcus** [reply]
> I don't think we need this in 2021, anyway it is possible to using `assets/inline`

I need this to instantiate workers hosted on a CDN.

**Reply by ShuZhong** [reply]
In latest webpack, that's a `workerPublicPath` to override `publicPath` in worker use.
we can set `publicPath: 'https://www.cdn.com` and `workerPublicPath: '/'` to solve CORS errors.

https://webpack.js.org/configuration/output/#outputworkerpublicpath

**Reply by CraigglesO** [reply]
This works! I had to do `publicPath: 'https://www.cdn.com` and `workerPublicPath: 'https://www.cdn.com` (They had to be the same url) myself. But other then that blown away and really happy this works. Supposedly it was around since `5.77.0` boy did i miss out. Appreciate you posting this. Saved me a lot of kB as well.

**Comment by alexander-akait** [comment]
How can I help?

Example required for
1. publicPath
2. inline worker fallback where worker doesn't supported

**Reply by AprilArcus** [reply]
I am working on a minimal repro for (2) today. Thanks for your help.

**Reply by chyzwar** [reply]
For us inline workers is main blocker for webpack5 adoptions. 
I am trying
```
{
    test: /\.m?js$/,
    resourceQuery: /inline/,
     type: 'asset/inline',
},
```
This need to be before/after babel loader ?

```js
 const worker = new Worker(new URL('./clusterBuilder.js?inline', import.meta.url));
```
I am still getting 
```
Failed to construct 'Worker': Script at 'https://local.com:8098/src_components_Map_components_Layers_components_PieClusterLayer_utils_clusterBuilder_js_inline.bdb12bf2ca2c37419f89.bundle.js' cannot be accessed from origin 'https://local.com:8097'.
```

Locally we have webpack dev server on https://local.com:8098 but app locally is running https://local.com:8097 via proxy that replace url on production page to local webpack. We need inline workers only for local dev. 

Our previusly used workaround that no longer work under webpack5
```js
/**
 * Provide workaround for cross origin workers
 * When running locally worker code is on webpack-dev-server
 * @see https://stackoverflow.com/questions/21913673/execute-web-worker-from-different-origin
 */
if (process.env.NODE_ENV === 'development') {
    const OriginalWorker = window.Worker;

    const PathWorker = (url: string | URL, opts?: WorkerOptions) => {
        const blob = new Blob(
            ['importScripts(' + JSON.stringify(url) + ')'],
            { type: 'text/javascript' },
        );
        return new OriginalWorker(URL.createObjectURL(blob), opts);
    };

    window.Worker = PathWorker as unknown as typeof Worker;
}

```

**Reply by chyzwar** [reply]
@alexander-akait  any chance you could provide example on how to do inline worker with webpack5?

**Reply by alexander-akait** [reply]
Do you need `base64` or `blob` inline?

**Reply by chyzwar** [reply]
I need to load worker when my page is running on different origin webpack dev server.  I will try to create simple example later today/tomorrow.

**Reply by alexander-akait** [reply]
Feel free to ping me

**Reply by chyzwar** [reply]
@alexander-akait @see https://github.com/chyzwar/webpack-inline-woker this should illustrate the issue.

**Reply by zenotsai** [reply]
I'm having the same problem upgrading webpack5 @chyzwar @alexander-akait

**Reply by alexander-akait** [reply]
@chyzwar solved, to be honestly it is not related to inline loader

**Reply by alexander-akait** [reply]
@chyzwar 

**index.html**

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Parent Page</title>
</head>
<body>
    <h1>Parent application</h1>
    <div id="root"></div>
    <script src="http://localhost:8080/main.js"></script>
</body>
</html>
```

**package.json**

```json
{
  "name": "@project/test-inline-workers",
  "version": "0.1.0",
  "private": true,
  "engines": {
    "node": "16.x.x",
    "yarn": "1.x.x"
  },
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-refresh": "^0.10.0"
  },
  "devDependencies": {
    "@babel/core": "^7.15.8",
    "@pmmmwh/react-refresh-webpack-plugin": "^0.5.1",
    "@types/webpack-manifest-plugin": "^3.0.5",
    "@types/node": "^16.11.1",
    "@types/react": "^17.0.30",
    "@types/react-dom": "^17.0.9",
    "babel-loader": "^8.2.2",
    "clean-webpack-plugin": "^4.0.0",
    "ts-node": "^10.3.0",
    "http-server": "^14.0.0",
    "source-map-loader": "^3.0.0",
    "ts-loader": "^9.2.6",
    "typescript": "^4.4.4",
    "webpack-manifest-plugin": "^4.0.2",
    "webpack": "^5.59.0",
    "webpack-cli": "^4.9.0",
    "webpack-dev-server": "^4.3.1"
  },
  "scripts": {
    "start:webpack": "webpack serve --config webpack.dev.ts",
    "start:server": "http-server . --proxy http://127.0.0.1:4000",
    "start": "yarn start:webpack & yarn start:server"
  },
  "browserslist": {
    "production": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

**webpack.config.js**

```js
import ReactRefreshWebpackPlugin from "@pmmmwh/react-refresh-webpack-plugin";
import type {Configuration} from "webpack";
import {ProvidePlugin, HotModuleReplacementPlugin} from "webpack";
import {resolve} from "path";
import { WebpackManifestPlugin } from 'webpack-manifest-plugin';

interface DevServerConfiguration {
  devServer: {
    port?: number;
    host?: string;
    allowedHosts?: string;
    historyApiFallback?: boolean;
    client?: {
      webSocketURL: {
        hostname: string;
        pathname: string
        port: number;
      }
    }
    headers?: {
      [key: string]: string;
    }
  }
}

const config: Configuration & DevServerConfiguration= {
  mode: "development",
  devtool: "inline-source-map",
  context: resolve(__dirname),
  entry: {
    main: resolve("./src/index.tsx"),
  },
  output: {
    path: resolve("build"),
    library: 'library', 
    crossOriginLoading: "anonymous",  
    uniqueName: 'library',
    umdNamedDefine: true,
    libraryTarget: 'umd',
  },
  resolve: {
    extensions: [".ts", ".tsx", ".js"],
  },
  target: "web",
  devServer: {
    port: 4000,
    host: "127.0.0.1",
    allowedHosts: 'all',
    historyApiFallback: true,
  },
  optimization: {
    runtimeChunk: true,
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        use: ["source-map-loader"],
        enforce: "pre",
      },
      {
        test: /(\.tsx|\.ts)$/,
        include: [
          resolve(__dirname, "./src"),
        ],
        use: [
          { 
            loader: "babel-loader", 
            options: {plugins: ["react-refresh/babel"]}
          },
          {
            loader: "ts-loader",
            options: {
              transpileOnly: true,
              projectReferences: true,
              experimentalFileCaching: true,
              configFile: "tsconfig.json",
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new WebpackManifestPlugin({}),
    new ProvidePlugin({
      React:     "react",
      ReactDOM:  "react-dom",
    }),
    new ReactRefreshWebpackPlugin(),
  ],
};

export default config;
```

**Reply by alexander-akait** [reply]
Shorty - just proxy dev server, it is some weird, actually `http-server` in your case before `webpack-dev-server`, so it will be right to proxy all request to `webpack-dev-server` and due `changeOrigin` (the option used inside `http-server`) you will get the valid origin and you don't need `CORS` in this way

**Comment by CraigglesO** [comment]
[Here is a good explanation on the problem and potential solution.](https://stackoverflow.com/questions/58098143/why-are-cross-origin-workers-blocked-and-why-is-the-workaround-ok?noredirect=1&lq=1)

edit -
@alexander-akait  - is it possible to create a loader for such a use case? Personally, I'd rather the scripts live separately, and the app fetch the workers using CORS.

**Reply by alexander-akait** [reply]
@CraigglesO You can use multi compiler mode with `target: 'webworker'` and load them from your application using absolve URL (proxy for local development)

**Comment by alexander-akait** [comment]
I see two things here:
- problem with dev server
- old browser support

Yes, inline loader can solve both them, but do you real need support old browsers? Anyway I provide two solutions here:
1. for dev server, just proxy - https://github.com/webpack/webpack/discussions/14198#discussioncomment-1514036
2. inline loader - 

**webpack.config.js**

```js
// ...
{
  resourceQuery: /raw/,
  // Note you can use `asset/inline` for very old browser supports, but you will increase bundle size
  type: 'asset/resource',
  use: [
    {
      loader: "babel-loader",
      options: {}
    },
    // Canb e removed if you don't have ts
    {
      loader: "ts-loader",
      options: {}
     }
  ]
} 
// ...
```

in source code just crreate

```ts
const createWorker = () => {
  if (process.env.NODE_ENV === 'development') {
    const data = new URL('./worker?raw', import.meta.url);
    
    // Can polyfill it, if you don't want, use something old to download file
    fetch(data)
        .then(response => response.blob())
        .then((blob) => {
          const objectURL = URL.createObjectURL(blob);
          
          new Worker(objectURL);

          URL.revokeObjectURL(objectURL);
    })
  } else {
    new Worker(new URL('./worker', import.meta.url))
  }
}
```

Technique from above can be improved for production usage, for example:
1. Using `data URI` for very old browser (note you will increase bundle size)
2. Creating data URI from fetched source and pass it to `new Worker`, if blob is not supported
3. Using `const isDevEnvOrOldBrowser = process.env.NODE_ENV === 'development' || typeof Worker === "undefined";`

Also you can always use `multi compiler` mode and bundle worker with `target: 'webworker'`.

Anyway feel free to feedback

**Reply by AprilArcus** [reply]
> do you real need support old browsers?

Yes, we have high-value clients who still rely on IE11. We're working on them, but they are entrenched and Microsoft is still supporting IE11 until next summer.

**Reply by AprilArcus** [reply]
I think we can solve this by emitting a stub worker that calls `importScript`

**Reply by alexander-akait** [reply]
@AprilArcus Why do not use `blob`?

**Reply by AprilArcus** [reply]
We serve our bundle from a CDN and it runs afoul of the same origin policy.

**Reply by alexander-akait** [reply]
In this case you can use `target: 'webworker'` and multi compiler mode

**Reply by zlk89** [reply]
I tried this script, but it seems webpack didn't transpile the code for me.

**Reply by Akimotorakiyu** [reply]
not work for me.
<img width="344" alt="image" src="https://user-images.githubusercontent.com/31165554/202625058-fda5145c-6890-4421-bc2e-798271d1ec3b.png">

**Comment by piotr-oles** [comment]
In case someone find it interesting, I found another solution to the CORS issue: https://github.com/webpack/webpack/discussions/14648#discussioncomment-1589272
